### DISCUSSION TITLE
Tree组件，无法区分“放置到内部”与“放置到第一个子节点之前”

### DISCUSSION THREAD

**Discussion by DViridescent** [ant-design/ant-design#51772] [Q&A]
## 背景

根据官网上的样例，可以在`onDrop`方法中，根据`info`来判断放置在某个节点的前、后、或者内部，如果放置在内部，则可以自行决定放置位置（样例代码选择了放置在内部的第一位）

我想让放置于节点内部时，放置到内部的末尾（而非第一位），然后遇到了一个问题。

## 问题

例如一个这样的树：

- 0-0
  - 0-0-0
  - 0-0-1

如果我拖动 `0-0-1`到`0-0-0`偏上的位置，组件会认为正在拖动到0-0的内部，而不是拖动到0-0-0的前面（此时`onDrop`的`info.dropToGap=false`）。

于是以下两个行为在`onDrag`方法中得到相同的参数
- 把一个节点放置到0-0内部
- 把一个节点放置到0-0-0前面

但我需要它们有不同的结果
- 放置到0-0内部时，我希望节点出现在0-0子节点列表的末尾
- 放置到0-0-0前面时，我希望节点出现在0-0-0子节点列表的第一位

类似的讨论：#31767

---

## Background

According to the sample on the official website, you can determine whether to place an item before, after, or inside a node using the `info` parameter in the `onDrop` method. If it's placed inside, you can choose the position yourself (the sample code selects the first position inside).

I want to place the item at the end inside a node, rather than at the first position, but I've encountered a problem.

## Issue

Consider a tree like this:

- 0-0
  - 0-0-0
  - 0-0-1

If I drag `0-0-1` to a position slightly above `0-0-0`, the component thinks it's being dragged inside `0-0` rather than in front of `0-0-0` (at this point, `info.dropToGap=false` in `onDrop`).

As a result, the following two actions receive the same parameters in the `onDrag` method:
- Placing a node inside `0-0`
- Placing a node in front of `0-0-0`

But I need them to have different outcomes:
- When placed inside `0-0`, I want the node to appear at the end of `0-0`'s child list.
- When placed in front of `0-0-0`, I want the node to appear at the beginning of `0-0-0`'s child list.

similar discussion: #31767

**Comment by petercat-assistant** [comment]
根据你的描述，问题在于无法区分将节点放置在父节点的内部还是放置在某个子节点之前。这种情况可能是由于 `onDrop` 方法中的 `info.dropToGap` 参数未能正确反映出用户的意图。

要解决这个问题，你可以尝试以下方法：

1. **自定义逻辑**：在 `onDrop` 方法中，除了使用 `info.dropToGap`，你还可以结合其他信息来判断用户的意图。例如，使用鼠标拖动的具体位置（如 Y 轴的偏移量）来判断是放置在某个子节点之前还是放置在父节点的内部。

2. **修改数据结构**：在拖动时，临时修改树的数据结构，以便更好地反映出拖动的意图。例如，可以在拖动开始时记录下拖动的节点和目标节点的关系。

3. **使用其他事件**：除了 `onDrop`，你还可以结合 `onDragOver` 或 `onDragEnter` 等事件来获取更多的上下文信息，从而更准确地判断用户的意图。

4. **参考类似讨论**：你提到的类似讨论 #31767 可能包含有用的解决方案或思路，可以查看该讨论以获取更多信息。

如果这些方法仍然无法解决问题，建议查看 ant-design 的文档或社区讨论，寻找其他开发者的经验和建议。

如果你有进一步的问题或需要更多帮助，请在评论中描述你的问题并 @petercat-assistant 以便我能继续帮助你。

**Comment by DViridescent** [comment]
写问题描述的时候突然灵光一现，可以通过在`onDragEnter`中存储一下当前鼠标hover的状态，然后在Drop的时候根据此时鼠标的位置是否在0-0-0上面来决定是否要执行移动到0-0-0之前的逻辑。

就是不知道有没有更好的方式？目前看上去不太优雅。
